mlflow/store/sqlalchemy_store.py

```diff
@@ -248,102 +278,104 @@ def _check_run_is_deleted(self, run):
-        run.lifecycle_stage = LifecycleStage.ACTIVE
-        self._save_to_db(run)
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_id, session=session)
+            self._check_run_is_deleted(run)
+            run.lifecycle_stage = LifecycleStage.ACTIVE
+            self._save_to_db(objs=run, session=session)
 
     def delete_run(self, run_id):
-        run = self._get_run(run_id)
-        self._check_run_is_active(run)
-        run.lifecycle_stage = LifecycleStage.DELETED
-        self._save_to_db(run)
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_id, session=session)
+            self._check_run_is_active(run)
+            run.lifecycle_stage = LifecycleStage.DELETED
+            self._save_to_db(objs=run, session=session)
 
     def log_metric(self, run_uuid, metric):
-        run = self._get_run(run_uuid)
-        self._check_run_is_active(run)
-        try:
-            # This will check for various integrity checks for metrics table.
-            # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
-            self._get_or_create(SqlMetric, run_uuid=run_uuid, key=metric.key,
-                                value=metric.value, timestamp=metric.timestamp)
-        except sqlalchemy.exc.IntegrityError as ie:
-            # Querying metrics from run entails pushing the query down to DB layer.
-            # Hence the rollback.
-            self.session.rollback()
-            existing_metric = [m for m in run.metrics
-                               if m.key == metric.key and m.timestamp == metric.timestamp]
-            if len(existing_metric) == 0:
-                raise MlflowException("Log metric request failed for run ID={}. Attempted to log"
-                                      " metric={}. Error={}".format(run_uuid,
-                                                                    (metric.key, metric.value),
-                                                                    str(ie)))
-            else:
-                m = existing_metric[0]
-                raise MlflowException('Metric={} must be unique. Metric already logged value {} '
-                                      'at {}'.format(metric, m.value, m.timestamp),
-                                      INVALID_PARAMETER_VALUE)
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_uuid, session=session)
+            self._check_run_is_active(run)
+            try:
+                # This will check for various integrity checks for metrics table.
+                # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
+                self._get_or_create(model=SqlMetric, run_uuid=run_uuid, key=metric.key,
+                                    value=metric.value, timestamp=metric.timestamp, session=session)
+            except sqlalchemy.exc.IntegrityError as ie:
```

Comment:
Is this logic gonna break down? If you share the `session` here -- `_get_or_create` will throw and then for the remainder of the function you will use the same session. If I am not mistake, that is invalid because SQLAlchemy requires that you `rollback` a session before performing new operations. I suspect, you might have to use a new session.

Also here -- the logic might conflate with multiple recorded metrics with different values for same timestamp.

Reply:
Discussed offline - this logic would indeed break down. Solved the problem by adding an explicit `commit()` statement before the `except` statement and an explicit `rollback()` after the `except` statement with substantial explanatory comments.